controller-server/src/main/java/com/yahoo/vespa/hosted/controller/notification/NotificationContentFormatter.java

```diff
@@ -0,0 +1,209 @@
+                            String.format("clusters,%s", clusterId.value()))
+                    .build();
+            return new NotificationContent(n, type, message, uri);
+        } catch (URISyntaxException e) {
+            throw new IllegalArgumentException(e);
+        }
+    }
+
+    private URI jobLink(NotificationSource source) {
+        var application = expect(source.application());
+        var instance = expect(source.instance());
+        var jobType = expect(source.jobType());
+        var runNumber = source.runNumber().orElseThrow(MissingOptionalException::new);
+        var applicationId = ApplicationId.from(source.tenant(), application, instance);
+        Function<Environment, URI> link = (Environment env) -> zoneRegistry.dashboardUrl(new RunId(applicationId, jobType, runNumber));
+        var environment = jobType.zone().environment();
+        switch (environment) {
+            case dev:
+            case perf:
+                return link.apply(environment);
+            default:
+                return link.apply(Environment.prod);
+        }
+    }
+
+    private String jobText(NotificationSource source) {
+        var jobType = expect(source.jobType());
+        var zone = jobType.zone();
+        var runNumber = source.runNumber().orElseThrow(MissingOptionalException::new);
+        switch (zone.environment().value()) {
+            case "production":
+                return Text.format("Deployment job #%d to %s", runNumber, zone.region());
+            case "test":
+                return Text.format("Test job #%d to %s", runNumber, zone.region());
+            case "dev":
+            case "perf":
+                return Text.format("Deployment job #%d to %s.%s", runNumber, zone.environment().value(), zone.region().value());
+        }
+        switch (jobType.jobName()) {
+            case "system-test":
+            case "staging-test":
+        }
+        return Text.format("%s #%d", jobType.jobName(), runNumber);
+    }
+
+    private String levelText(Notification.Level level, int count) {
+        switch (level) {
+            case error:
+                return "failed";
+            case warning:
+                return count > 1 ? Text.format("%d warnings", count) : "a warning";
+            default:
+                return count > 1 ? Text.format("%d messages", count) : "a message";
+        }
+    }
+
+    private String clusterInfo(NotificationSource source) {
+        var application = expect(source.application());
+        var instance = expect(source.instance());
+        var zone = expect(source.zoneId());
+        var clusterId = expect(source.clusterId());
+        return Text.format("Cluster %s in %s.%s for %s.%s",
+                clusterId.value(),
+                zone.environment(), zone.region(),
+                application, instance);
+    }
+
+    private URI url(Notification n) {
+        var source = n.source();
+        if (source.application().isPresent()) {
```

Comment:
Nit: You can avoid the excessive nesting if you restructure a bit, i.e:

```
if (source.application().isEmpty()) { return ... }
if (source.instance().isEmpty()) { return ... }
if (source.jobType().isEmpty() || source.runNumber().isEmpy() { return ... }
return ...
```

Reply:
I removed the entire thing, it was only needed for unknown Notifications, and that is something that should not really happen.